### PR TITLE
TOC: replace "Table of Contents" with book title

### LINKED ISSUE #9042: Chapter Titles in the Alt Status Bar
**Context:**
- Chapter titles are useful for acknowledging the reading context especially when reading several books and switching between them
- Bottom status bar has **chapter titles** but it is generally crowded with several items which can only be added to the bottom bar. Not much space is available for chapter titles.

**Proposal:**
If it is not too much work, I think having the option to have **Chapter Titles** in the **Alt Status Bar** will be useful

**Benefits:**
- This option can free up precious bottom bar space
- Some chapter titles can be very long so they may need wider space to fit, **Alt status bar** is ideal for this
- It will look more natural because printed books generally have **chapter title** and **page number** in the header
- BONUS: If bottom bar is hidden while using this feature, screen will have a clean "printed book" feeling

chapter title

### PR DESCRIPTION
**Pull request by zwim**
If a boot title is available replace the generic "Table of Contents" with the book title. (Is shortened, when too title is to long).

In the following image is a collage of three books. The first has not title, the scond one a very long title and the last is shown as it is.

This PR could fix #9042.

grafik

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/9058)

### CODE REVIEW

**Comment by hius07** [comment]
Don't you want to show the full title, without truncating, in multilines?
`title_multilines` TitleBar property.

**Comment by zwim** [comment]
I would go with shortening. Some thoughts:
1. Most off the time the title would be short, anyway.
2. A multiline title might be broken on wrong positions, then it looks clumsy.
3. A multiline title takes away room.
4. A single line looks clean.

What do you think?


Pinging also @offset-torque @poire-z

File: frontend/apps/reader/modules/readertoc.lua [hunk 1]
```diff
@@ -753,8 +753,14 @@ function ReaderToc:onShowToc()
     --       because we inflate the state Button's width on the left, mainly to give it a larger tap zone.
     --       This yields *slightly* better alignment between state & mandatory (in terms of effective margins).
     local button_size = self.expand_button:getSize()
+
+    -- Get book title if available, else use generic "Table of Contents"
+    local doc_info = self.ui.document:getProps()
+    local title = doc_info and doc_info.title ~= "" and doc_info.title or _("Table of Contents")
+    title = title:gsub(" ", "\xC2\xA0") -- replace space with no-break-space
```
Comment (NiLuJe) [inline frontend/apps/reader/modules/readertoc.lua:760]: Curious cat is curious: what's with standard spaces? :}

**Comment by poire-z** [comment]
Not fond of the idea.
Why for Table of contents, and not for Book Information, Book map, Page browser, Reading statistics ?
I prefer to not just see the title: it's available to me via another gesture that shows me Book information.

And that may be just me, but I don't take care of my metadata - but I make clean filenames. In CoverBrowser, I can just "Ignore metadata" when they are crap - and never see the crap (except in a single place: Calendar view :/).
So, I don't need to see the crap at the top of Table of content.

**Comment by NiLuJe** [comment]
I'd possibly shove it somewhere in a menu entry label somewhere, but, then again, not terribly fond of the idea on a basic level (I mean, you probably have bigger issues if you're wondering exactly which book you currently have open. In which case you can do the equivalent of flipping the book around by hitting the power button, as most users should either have the cover or the metadata shown somewhere in there).

**Comment by zwim** [comment]
Ok, it seems common sense is, not to use the book title in the title line of toc, book map ...

So closing this.

**Comment by offset-torque** [comment]
Firstly I don't understand the resistance to showing users the name of their book in an easily accessible and related place.

That **Table of contents** title is totally redundant and can be replaced. Does anyone really think it is necessary ? Do you really think it is serving a purpose ? 

I can't imagine a user:
- opening the menu
- tapping on **Table of contents**
- and while looking at chapter names wondering **"Where the hell I am ? What is this window ?"**

You have to tap on **Table of contents** to open **Table of contents**, can it be more explicit ? Every single user opening the **Table of contents** knows that he is in **Table of contents**. It doesn't need further explanation.

---
@poire-z 
> **I prefer to not just see the title: it's available to me via another gesture that shows me Book information.**

Then don't look at it and continue using your gestures and let new users enjoy the convenience. I bet nearly all of the KOReader users know how to open the **Table of contents** and they are regularly opening it. How many of the users do you think know the gesture for **Book information** (is there even a default gesture for it). How I wish there was analytics and telemetry so I could back up my UI/UX claims.

**Two passages from the future user guide:**

If you want to check the title of the book you are in, you can:
- **Proposal:** Open the table of contents (➡️⬅️ multiswipe)
- **Currently:** Open the top menu. Tap on Hamburger menu icon. Tap on Book Information. (One swipe+2 taps and user now has to memorize the location of the item unlike the **ToC** option which he already uses OR set and memorize another gesture just to see the title)

---
@NiLuJe 
> **you probably have bigger issues if you're wondering exactly which book you currently have open**

That's a short-sighted take. Not everyone is reading just the latest bestsellers. Academic reading requires reading books with similar topics at the same time when you are researching a subject. 

Here, example chapter names from two books I am currently reading:
- **Book 1:** Active Inference and the Primacy of the ‘I Can’
- **Book 2:** Active Inference as a Unified Theory of Sentient Be­hav­ior

These two books have different approaches to the subject, so knowing which book I am in helps with setting the cognitive context.


> **by hitting the power button, as most users should either have the cover or the metadata shown somewhere in there**

I hope this is a joke. Imagine the user guide:
- If you want to check the title of the book you are in, just turn off your e-reader

:)))

---
@zwim 
I would have said the full title but considering the points you provided, truncated makes more sense. I checked my reading folder and majority of the book titles are between 3-6 words long so even when it is truncated, main part of the title should be visible and enough to identify the book.

**Comment by zwim** [comment]
Reopened, cause there seems some need for discussion.

**Comment by NiLuJe** [comment]
I'm more puzzled about the whole concept, the implementation is fine, as far as I'm concerned ;).

There may be a case to be made to unify that w/ the other widgets @poire-z mentioned, though (although I'm not familiar with how they look off the top of my head). The ToC feels simple enough and a somewhat natural place for it, though.

**Comment by poire-z** [comment]
> If you want to check the title of the book you are in, you can:
    Proposal: Open the table of contents (➡️⬅️ multiswipe) (hint: book title is father of all chapter titles)

Let's go further:

> If you want to check the author of the book you are in, you can:
    Proposal: Open the Book map (hint: map comes from the author brain)

> If you want to check the series of the book you are in, you can:
    Proposal: Open the Page browser (hint: "series of pages")

> If you want to check the keywords of the book you are in, you can:
    Proposal: Open the Bookmarks (hint: "keywords are kinda marks for the book")

> If you want to check the year of the book you are in, you can:
    Proposal: Open the Skim widget (a year is small, we'll find some place in this widget for it)


Seriously, there are more convenient way to see all this: show History: you get what you decided for, title, authors, cover, series combined with titles - and you get your chronology of previous books.
And show Book information: you get all of the metadata.
Don't mislead users with half-bad tricks in a user guide, tell them the good truth and right way to get to stuff.
(And I hope this won't have to go into a User guide: there are various ways to get to the metadata bit you want, let the users know about the various options, and let them decide what works for them, they're not that stupid that we need to tell them how to see the title and authors and years of the book they are reading...)

The "book title is the father of all chapter titles" is something I would buy and accept, if the book title came from the same source as the chapter titles. (It's possible one of the ToC formats supported by crengine has a "level 0" title, that we may discard as it does not fit in our model of "a ToC has multiple level 1 chapters" - that would fit better as the title for this ToC widget - as it comes from the same source.)

As I mentionned above, the book title might not be correct (and we have no way from inside KOReader to correct it).
I don't want to have to ignore it: show me the Table of contents as defined by the book, don't assume the title of the book is just right for the hat on the ToC.
I want KOReader to be consistent and not a patchwork of various ideas and dubious convenience shortcuts.
I'm not particular fond of the ToC/Bookmarks/... title - for my usage, I would rather not have them and have the full height for the content. But it's there, it's consistent, it's useful for some people.

**Comment by NiLuJe** [comment]
Devil's advocate: metadata being crap is more of a user's problem than a KOReader problem, though.

(I mean, the only thing I do before sending books is, err, fix the basic metadata to make it consistent with how I like it, because it also happens to be used to dictate the filesystem layout how I like it).

**Comment by NiLuJe** [comment]
TL;DR: I don't care any other way, because I don't necessarily see this as terribly useful, but I also don't see what would warrant shooting down a fairly innocuous implementation this hard ;).

**Comment by zwim** [comment]
All in all, seeing the title of the "Romeo & Juliet" instead "Table of contents", seems to give the dialog a more personal touch. 

TOC is really generic, and I know that I see the TOC (because I opened it). The generic title is actually only noise on the top, eating space (but dropping it, is no option), so why not replace it.

The other dialogs: `bookmap` is something special, here I am happy to see how this feature is called. `pagebrowser` is also special, as it can be called from `bookmap`. IMHO it is important to keep the other names to distinguish, where we are in. 

Back to the title of the book and the table of contents. If I see the title and the content, it is apparently where I am in, so no need for "Table of context" as a dialog title.

**Comment by hius07** [comment]
> eating space

Maybe it's not a bad idea to have a setting to hide the header.

**Comment by Frenzie** [comment]
@NiLuJe 
> I'd propose allowing a fall back to the filename as a compromise, but it's often as useless (if not more) than the metadata (before user intervention). (And would be far worse than nothing in my book anyway, FWIW).

The filename is as useless if you don't change it, but the filename is something you quickly adjust while downloading. E.g., if I download the document 234asdfs343.pdf from JSTOR I save as `Surname 2022 - Maybe title 234asdfs343.pdf`. Changing filenames is virtually indistinguishable from sorting files in folders. Changing metadata is something significantly more involved, not something I'd wish on my worst enemies.

@hius07 
> Maybe it's not a bad idea to have a setting to hide the header.

But not the close button. ;-)

**Comment by hius07** [comment]
> But not the close button. ;-)

It can be closed with double-finger swipe west and one-finger swipe down (the latter needs some fixing).

**Comment by offset-torque** [comment]
If I understood correctly, the only objection can be summarized as:
**"Book title/metadata can be messy and showing it at the top of the ToC might be ugly"**

I can understand this concern because some downloaded files are really atrocious. But it is like saying **"I can't cook you a beautiful dinner so I won't give you any food"** to your hungry friend.

Maybe we can give the user the option to choose, so everyone can be happy:
" :ballot_box_with_check:  Show book title in Toc"

And +1 for "fallback to file name".

---

@poire-z 

> Don't mislead users with half-bad tricks in a user guide, tell them the good truth and right way to get to stuff.

Every part of that guide is approved by developers. I am not writing a novel and publishing it independently.

> And I hope this won't have to go into a User guide: there are various ways to get to the metadata bit you want, let the users know about the various options, and let them decide what works for them, they're not that stupid

I am preparing the user guide for the average user and my motto is **"Good enough"**. My intention is not documenting every single feature of KOReader. KOReader is not like Blender where skill levels are concentrated at the long tail. We have old Dorothy who just want to read her Bible on her Kindle or little Emily with her vampire stories and they don't give a crap about intricacies of stylesheets or publisher metadata. E-readers are not hi-tech items anymore, they are now sold in Walmart for the price of a dinner.

And remember that there is also a wiki which can also be filled with any levels of details for the curious users. If someone wants to add detailed info they are welcome to fill up the wiki. Here I am trying to prevent average user from searching google and diving into forums just to find some basic usage info. People who can write Calibre plugins to organize their library can skip the user guide and go directly to wiki.

**Users are not stupid but they are hardly at the skill level you assume, so I have to simplify things in the guide.**

Here is a study about computer skill levels of the population:
- Below Level 1 - “Delete this email message”
- Level 1 - “Find all emails from John Smith.”
- Level 2 - “You want to find a sustainability-related document that was sent to you by John Smith in October last year.”
- Level 3 - “You want to know what percentage of the emails sent by John Smith last month were about sustainability.”

Relevant quote from the study:

> The main point I want to make is that you, dear reader, are almost certainly in the top category of computer skills, level 3. Overall, people with strong technology skills make up a 5–8% sliver of their country’s population, whatever rich country they may be coming from. Look at skill Level 3. What’s important is to remember that 95% of the population in the United States (93% in Northern Europe; 92% in rich Asia) cannot do these things.


computer-skill-levels-by-country

**Comment by Frenzie** [comment]
> If I understood correctly, the only objection can be summarized as:
"Book title/metadata can be messy and showing it at the top of the ToC might be ugly"

Well no, if I have an objection to the PR it would be that it's inconsistent. The metadata thing is just a red herring.

**Comment by mergen3107** [comment]
Regarding the "it is obvious that this is ToC" - what if the user opened it by gesture? and accidentally? Of course, it is "obvious" that this a ToC in front of you - but only given there are meaningful chapter names. Sometimes each chapter is labeled "***" or "-", then the user will be confused. Not like this is a big problem, but still a confusion in the UI.

**Comment by Frenzie** [comment]
But to be fair, as long as you can close it using the obvious X without having to know swipe down, that doesn't really matter. Being able to get out of it is the most important. Also it seems like a rather unlikely scenario based on my n=1.

I do have a lot of books that come with a big H1 header title and copyright page on page 1 or 2, so in the example from the OP they'd be:

# Romeo and Juliet
Romeo and Juliet .... 2

In fact you can even see in the screenshot that Romeo and Juliet is one of these very books! Except for some reason it's called "Title" in the ToC rather than "Romeo and Juliet".

**Comment by zwim** [comment]
I would go with @offset-torque and let the user decide what to see. We already have a settings menu (which is not crowded) so:

grafik

**Comment by NiLuJe** [comment]
As per the original question, my vote would be for one like, truncated, with the full entry on hold like the chapter labels in the toc.

**Comment by Frenzie** [comment]
Well, I'm not really objecting as such. I'm fairly ambivalent. I am however also definitely not positive. ;-P

**Comment by poire-z** [comment]
> Devil's advocate: metadata being crap is more of a user's problem than a KOReader problem, though (i.e., garbage in, garbage out; which is a fine approach to take as far as metadata is concerned, because it's fairly often pretty garbage indeed ;p. So if you decide to not show it simply on the principle of it being crap, taking that to the extreme would imply not showing any of it ^^).
(I mean, the only thing I do before sending books is, well, fix the basic metadata to make it consistent with how I like it, because it also happens to be used to dictate the filesystem layout how I like it. But, more importantly, it also ensures I can actually find stuff properly, be it on device or in my local library).

So, letting you know of something you (calibre users) may not know: it's neither a user nor a KOReader problem. It's just not a problem at all (except in Statistics plugin, and with this PR...)
Garbage in, garbage out, garbage hidden.
KOReader is good, gentle, the most kind, it doesn't impose on you, it does not boast, it is not proud. It does not dishonor others (by forcing them to use calibre), it is not self-seeking, it is not easily angered, it keeps no record of wrongs. KOReader does not delight in evil but rejoices with the truth. It always protects, always trusts, always hopes, always perseveres!

The only places we see this possible-crap metadata are:
- book  information (but ok, you get to see what is there)
- cover browser : where you can quickly "Ignore metadata" for a book and be done witht it, you won't see then in File browser nor in history.
- this ToC widget getting the title metadata as the widget title (that, ok, thank you, I can just disable globally - and only you users of the feature will see the possible crap, and may be you'll end up disabling it globally too as you can't do it per book :) 
- Statistics plugin & calendar view (title+author being used as a key for stats, and being used for display - I hope we can find a solution for this one, but it may need a stat DB update...)

So, garbage metadata is not really a problem with KOReader, and it doesn't require you to have to clean your files before they get usable - and to have to use some specific software to manage/transfer them. Which is a big feature of KOReader for me, that I'll fight to keep.
(I'm grateful all the metadata broswing/searching has been kept into the calibre plugin and that I can safely ignore it.)

File: frontend/ui/elements/reader_menu_order.lua [hunk 2]
```diff
@@ -31,6 +31,7 @@ local order = {
     },
     navi_settings = {
         "toc_alt_toc",
+        "toc_show_book_title",
```
Comment (poire-z) [inline frontend/ui/elements/reader_menu_order.lua:34]: I think it should be below, above toc_items_per_page, as it's about the look/display of the ToC.
Here, toc_alt_toc and toc_ticks_level_ignore are about the ToC content

**Comment by poire-z** [comment]
> But it is like saying "I can't cook you a beautiful dinner so I won't give you any food" to your hungry friend.

Rather "I can't cook you a beautiful dinner but I will give you bread and cheese, as I give everyday to any visitor, friend or foe".

But ok, you are 3 that want this, and it's not the default. (just another wee ridiculous setting), so go ahead...

I still think it's really not a great idea.
"Table of contents" just work in all contexts. It's common, it's classic, it's fixed, it's safe, it's consistent.

Book title will:
- for 50% of books be small enough to not be truncated - and it will look nice.
- for 30% more of books it will be truncated (`hamburger The wonderful adventures of... the close button`) and will not look as nice, and will add some noise/weight at the top that will catch your eyes
- for 10%, you'll get a bad book title: live with it
- for 10% (many PDFs), you'll get NO ToC at all: you'll get a screen with "123321312.pdf   x" at top, and " << < no item > >>" at bottom. It would look better with "Table of content : empty : no item".

You could say we don't care for the remaining 20% , it's just ToC after all, it won't hurt if there is crap. Except, we didn't have crap before. And it was just ToC after all, that you don't see often, so you didn't really need to have cherry at top.

But I know we won't say that, because KOReader perseveres, and won't leave crap be crap. So, we will work on it, and decide that when no ToC content, we will show "Table of contents" instead of the book title. And we'll discuss it, argue, add checks to this simple PR, and we will be overengineering this initially "fairly innocuous implementation".
While we were just fine with bread and chesse - but we were greedy for cherries and cake.

**Comment by zwim** [comment]
@poire-z Thanks for your exhaustive explanations. Undoubtedly you know more about the quirks and pitfalls as I do, and I honor your experience. I appreciate learning from your thoughts.

I also agree, that we shouldn't overengineer this whole thing.

A proposal, which might fit your concerns and give us honey and milk:
What about making the title `Table of contents` tappable?
1. Per default we start with `TOC`.
2. Tap: -> Show an the book information form the hamburger menu.
3. Tap and hold -> Toggle the default behavior to show the title instead of TOC. (Tap and hold again back to TOC) **per book **

Maybe 2 and 3 the other way round.

Andwe drop the additional settings entry.

**Comment by hius07** [comment]
> What about making the title `Table of contents` tappable?

It is easier (and consistently) to add the title left button for that.

**Comment by Frenzie** [comment]
Yes, what @hius07 said and *only* what @hius07  said is what I like.

**Comment by offset-torque** [comment]
Arguments in @poire-z's last message are logical. Also I forgot about the **History**. It is also a one gesture access to the book title. So considering these, I am reverting my position on the issue. Especially I don't want to spend dev effort considering the "overengineering" part for a simple thing like this.

I think **History** is widely used considering that we don't have a **HOME** screen like other readers. I can give a reminder to the users that they can find current book title in the **History** page. Maybe we can even highlight the current book in there like with a darker border or a symbol to make it stand out. (I know first book in the list is the current one but a visual indicator might be nice)

**Comment by zwim** [comment]
What about: Tapping on the info icon opens "Book information", a long tap toggles "Table of contents" and "Rome and Juliet" and saves the current state in the siedcar. (and all with additional 11 lines of code :) )

grafik

**Comment by offset-torque** [comment]
> Tapping on the info icon opens "Book information"

I think this is a natural and easier access to the **Book information** which is a good idea.

> a long tap toggles "Table of contents" and "Rome and Juliet" and saves the current state in the siedcar.

The problem is, **long-press on info button** is not a discoverable action. When a user presses on that button first time, they will register that function mentally and probably won't even try to long-press unless there is an indicator for alternate function (probably not possible in KOReader). Of course I can explain it in the user guide but current users probably won't notice. So I am not sure if this will worth the effort and added complexity. Once a feature is tucked deep in somewhere, its utility decreases proportionally to its discoverability. So this will be more like a non-existent feature providing only marginal benefit overall (unlike a **default** which doesn't expect anything from the user).

My main argument for this issue was for an **"easy access to current book title"** and **History** dialog can satisfy this condition it seems (familiar dialog, even has a default gesture) without the downsides mentioned. So I can say **Book title** issue looks solved for now.

**Comment by offset-torque** [comment]
Thanks for giving my idea a try **zwim** also others for contributing to the discussion. I appreciate the efforts :wave:

**Comment by zwim** [comment]
Despite the consensus here, I would like to keep that PR on my device.
What would be the standard way to let it survive OTA updates? 

(Tap on the top left `left-right-arrows` icon toggles "TOC" <-> "Title"; no hold action)
grafik

**Comment by poire-z** [comment]
> What would be the standard way to let it survive OTA updates?

There is no standard way :)
We could probably add a way to execute some user `koreader/monkeypatch.lua` at some point in the loading process so one could live-patch some KOReader classes code on startup, that could be helpful and simple for small overrides/additions.
But here, you are hacking a bit too much inside the onShowToc() method for it to be monkey-patched...
(For this kind of thing, I do it in my own patch that I apply manually after a manual install - that I have to often fix and update manually - I guess there's no real solution to have it done when using nightlies)

What could be done if you're fine without the buttons, and just using the book title, is just having in koreader master code a hidden/manual setting that users in the know could manually add:
```lua
title = _("Table of Contents")
if G_reader_setting:isTrue("use_book_title_as_toc_title") then -- (hidden setting not exposed in the UI)
    local doc_info = self.ui.document:getProps()
    if doc_info and doc_info.title ~= "" then title=doc_info.title end
end
```

**Comment by Frenzie** [comment]
> What would be the standard way to let it survive OTA updates?

I don't know about a standard way, but there are several kinds of patch-based approaches. My preference would probably go to something like this:

```
for file in $({ git diff --name-only; git diff --name-only origin/master; } | sort | uniq); do [[ -f "${file}" ]] && scp "${file}" "root@kobo:/mnt/onboard/.adds/koreader/${file}"; done
```

In that case you maintain a branch occasionally rebased on top of master with some patches.

PS I don't do any such thing. :-)

**Comment by zwim** [comment]
Thanks:

@poire_z: But are there any other hidden settings, or would this be the only one?

@Frenzie: I will go for the `git diff` version.

Just in case anyone is interested I will commit my branch here.

**Comment by poire-z** [comment]
> But are there any other hidden settings

There are a few, but they should stay a few, and be only micro-behaviour tweaks.

One I can think of that i added mostly for my own usage:
https://github.com/koreader/koreader/blob/4b0f7471f4a38c3371d01a8251d8f9f9f2937eb8/frontend/ui/bidi.lua#L76-L78
Some other one, where it feels it didn't deserve to be in a menu, but has helped with some Libra2 issues:
https://github.com/koreader/koreader/blob/4b0f7471f4a38c3371d01a8251d8f9f9f2937eb8/frontend/apps/reader/modules/readerrolling.lua#L678-L688

May be some of the ones I described at https://github.com/koreader/koreader/issues/2589#issuecomment-323927682 are still usable and hidden.

**Comment by poire-z** [comment]
> We could probably add a way to execute some user koreader/monkeypatch.lua at some point in the loading process so one could live-patch some KOReader classes code on startup, that could be helpful and simple for small overrides/additions.

Well, couldn't resist trying to do that :)
(I actually do this kind of thing to my Firefox, patching Firefox and some extensions' javascript to tweak their behaviour to my liking.)

If we add:
```diff
--- a/reader.lua
+++ b/reader.lua
@@ -257,4 +257,14 @@ local function retryLastFile()
 end

+-- Run any koreader monkey patch
+-- (We could have this be a directory and run all the files in this directory)
+local monkopatch = DataStorage:getDataDir() .. "/monkopatch.lua"
+if lfs.attributes(monkopatch, "mode") == "file" then
+    local ok, err = pcall(dofile, monkopatch)
+    if not ok then
+        require("logger").warn(err)
+    end
+end
+
 -- Start app
 local exit_code
```
@zwim (and likely minded friends) could use a `koreader/monkopatch.lua` with:
```lua
local Event = require("ui/event")
local UIManager = require("ui/uimanager")
local _ = require("gettext")

-- local InfoMessage = require("ui/widget/infomessage")
-- UIManager:show( InfoMessage:new{ text="patching ReaderToc" } )

local ReaderToc = require("apps/reader/modules/readertoc")
local onShowToc_orig = ReaderToc.onShowToc
ReaderToc.onShowToc = function(self)
    -- Run original code
    onShowToc_orig(self)
    -- See if the book has some title
    local doc_info = self.ui.document:getProps()
    if doc_info and doc_info.title ~= "" then
        -- Add some left icon, that will then be used after :init() below
        self.toc_menu.title_bar.left_icon = "info"
        -- Set book title as TOC widget title (or keep the original if it has been disabled)
        local title = _("Table of Contents")
        if not self.ui.doc_settings:isTrue("toc_hide_book_title") then
            title = doc_info.title
        end
        -- Re-init TitleBar
        self.toc_menu.title_bar:clear()
        self.toc_menu.title_bar.title = title
        self.toc_menu.title_bar:init()
        -- Add some handlers for the added left button
        function self.toc_menu:onLeftButtonHold()
            self.ui.doc_settings:toggle("toc_hide_book_title")
            self:onClose()
            UIManager:broadcastEvent(Event:new("ShowToc"))
        end
        function self.toc_menu:onLeftButtonTap()
            UIManager:broadcastEvent(Event:new("ShowBookInfo"))
        end
    end
end
```

**Comment by zwim** [comment]
Cool, I will try that, when I am at home.

**Comment by zwim** [comment]
Really cool, it works.
Shall I do a PR for the `reader.lua` changes. My current version looks like:
``` diff
diff --git a/reader.lua b/reader.lua
index 07120700..84da4ba2 100755
--- a/reader.lua
+++ b/reader.lua
@@ -256,6 +256,29 @@ local function retryLastFile()
     }
 end
 
+-- Run any koreader monkey patches directly in the monkopatches directory (flat, no subdir).
+-- (https://github.com/koreader/koreader/pull/9058#issuecomment-1120240470, thanks to @poire-z)
+local monkodir = DataStorage:getDataDir() .. "/monkopatches"
+if lfs.attributes(monkodir).mode == "directory" then
+    local logger = require("logger")
+    logger.info("Monkey patch directory found")
+    for entry in lfs.dir(monkodir) do
+        local monkopatch = monkodir .. "/" .. entry
+        if lfs.attributes(monkopatch).mode == "file" and monkopatch:match(".lua$") then
+            local ok, err = pcall(dofile, monkopatch)
+            if not ok then
+                -- Only developers (advanced users) will use this mechanism.
+                -- A warning on a patch failure after an OTA update will simplify troubleshooting.
+                logger.warn(err)
+                local InfoMessage = require("ui/widget/infomessage")
+                UIManager:show(InfoMessage:new{text=_("Error loading patch:\n") .. monkopatch})
+            else
+                logger.info("Monkey patch applied:" .. monkopatch)
+            end
+        end
+    end
+end
+
 -- Start app
 local exit_code
 if file then
```

**Comment by poire-z** [comment]
> Shall I do a PR for the reader.lua 

I'm fine with having that, dunno about @Frenzie and @NiLuJe .
Points to think about:
- at which point to include it in reader.lua (after sufficient stuff has been loaded, but then some stuff may already be in use, ie. you probably can't hack UIManager too much).
- the name of the directory, my `monkopatch` was half a joke :)
- and it it's really only for "monkey patching", or if it could have other uses

**Comment by Frenzie** [comment]
Note that there's already something similar for Android, in case it makes sense to unify this kind of thing (probably not, but you should at least check ;-) ).

**Comment by NiLuJe** [comment]
>     * at which point to include it in reader.lua (after sufficient stuff has been loaded, but then some stuff may already be in use, ie. you probably can't hack UIManager too much).

Probably right after the UIManager require.
Leaving Device & UIManager alone seems like a safe bet anyway ;).

>     * the name of the directory, my `monkopatch` was half a joke :)

KISS: `patches` ;o).

**Comment by NiLuJe** [comment]
> Note that there's already something similar for Android, in case it makes sense to unify this kind of thing (probably not, but you should at least check ;-) ).

That's handled in an Android-specific codepath *before* we even load reader.lua, so it'll probably have to be left alone ;).

**Comment by poire-z** [comment]
User patch available at https://gist.github.com/zwim/1edcd34ef8a59166f203d5ee8c08f7e3

**Comment by mergen3107** [comment]
@hius07 
Can you please suggest what variable to use if I want to have the book title from custom metadata? Is it even possible?

Here is an example. I got a PDF file with poor metadata. I don’t have a good metadata editor on PC, so I just put the PDF in Kindle with KOReader and edited its title with custom metadata. 

Now I would like to see this edited title shown in the TOC using this user patch, but I can only see the original title.

Thank you!

**Comment by hius07** [comment]
Try `self.ui.doc_props` in line 13.

**Comment by mergen3107** [comment]
Thank you very much @hius07 ! That works :)
